### PR TITLE
914 add building blocks to module

### LINKED ISSUE #914: Create "Add Building Blocks" Context Menu for Module
When right clicking on a module, the user should get a context menu "Add Building Blocks",

image

and by clicking it the user gets a list of all the building blocks that the module does not already have (since we should have only a single BB of each type per molecule), and can select which ones to add.

image

### PR DESCRIPTION
**Pull request by georgeDaskalakis**
closes #914 

@msevestre this is ready - but as I told you last week I have a few points on the code where I am not 100% sure it is the correct design. We can talk about this on a call too whenever you want.

### CODE REVIEW

File: src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+using OSPSuite.Core.Domain.Builder;
+using OSPSuite.Utility.Extensions;
+
+namespace MoBi.Core.Commands
+{
+   public class AddBuildingBlocksToModuleCommand : MoBiReversibleCommand, ISilentCommand
+   {
+      protected Module _existingModule;
+      protected Module _moduleWithNewBuildingBlocks;
+      public string ExistingModuleId { get; private set; }
```
Comment (msevestre) [inline src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs:16]: why are those guys public?

File: src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+         ModuleWithNewBuildingBlocksId = moduleWithNewBuildingBlocks.Id;
+         Silent = false;
+         Description = AppConstants.Commands.AddBuildingBlocksToModule(_existingModule.Name);
+      }
+
+      //ToDo: this should be refactored to use a generic method - ideally also reflection 
+      //to go through the class members
+      protected override void ExecuteWith(IMoBiContext context)
+      {
+         context.Register(_existingModule);
```
Comment (msevestre) [inline src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs:35]: ?why are you registering n the command/ They arlead exist right?

File: src/MoBi.UI/MoBi.UI.csproj [hunk 3]
```diff
@@ -57,6 +57,9 @@
       <AutoGen>True</AutoGen>
       <DependentUpon>Resources.resx</DependentUpon>
     </Compile>
+    <Compile Update="Views\AddBuildingBlocksToModuleView.cs">
```
Comment (msevestre) [inline src/MoBi.UI/MoBi.UI.csproj:60]: remove

File: src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+namespace MoBi.Core.Commands
+{
+   public class AddBuildingBlocksToModuleCommand : MoBiReversibleCommand, ISilentCommand
+   {
+      protected Module _existingModule;
+      protected Module _moduleWithNewBuildingBlocks;
+      public string ExistingModuleId { get; private set; }
+      public string ModuleWithNewBuildingBlocksId { get; private set; }
+
+      public AddBuildingBlocksToModuleCommand(Module existingModule, Module moduleWithNewBuildingBlocks)
```
Comment (msevestre) [inline src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs:19]: Why is the use case a module to a module and not a buildingblock to a module
I am having trouble seeing how this is supposed to work

**Review by msevestre — CHANGES_REQUESTED** [review]
lets' discuss

File: src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs (hunk 2 above)
Comment (georgeDaskalakis) [reply src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs]: removed

**Comment by georgeDaskalakis** [comment]
@msevestre I have changed the logic as we discussed ( moduleExtensions to handle the adding of BB, MacroCommand and then individual commands for adding the separate BBs and so on ). I am currently adjusting the unit tests, but please take a look on the logic in case something needs changing
. 
Since this is an intermediate state, I am leaving the PR as a draft.

File: src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs (hunk 1 above)
Comment (georgeDaskalakis) [reply src/MoBi.Core/Commands/AddBuildingBlocksToModuleCommand.cs]: made private

**Comment by georgeDaskalakis** [comment]
@msevestre I also added the unit tests. If it looks good for you, I will merge.

File: src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
+﻿using MoBi.Assets;
+using MoBi.Core.Exceptions;
+using OSPSuite.Core.Domain;
+using OSPSuite.Core.Domain.Builder;
+
+namespace MoBi.Core.Domain.Extensions
+{
+   public static class ModuleExtensions
```
Comment (msevestre) [inline src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs:8]: Should we move this to core?

File: src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+               module.PassiveTransports = null;
+               break;
+            case IEventGroupBuildingBlock eventGroup:
+               module.EventGroups = null;
+               break;
+            case IObserverBuildingBlock observer:
+               module.Observers = null;
+               break;
+            case ParameterStartValuesBuildingBlock parameterStartValues:
+               //module.RemoveParameterStartValueBlock(parameterStartValues);
```
Comment (msevestre) [inline src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs:68]: What's with this?

**Review by msevestre — COMMENTED** [review]
Looks good. Just a few questions

File: src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs (hunk 5 above)
Comment (georgeDaskalakis) [reply src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs:8]: done

File: src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs (hunk 6 above)
Comment (georgeDaskalakis) [reply src/MoBi.Core/Domain/Extensions/ModuleExtensions.cs:68]: Is the new functionality in this comit:
https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/pull/1966

was commented out for the solution to build. MOved to Core now anyways

**Comment by georgeDaskalakis** [comment]
@msevestre I have moved the functionalities to Core, in this PR:

https://github.com/Open-Systems-Pharmacology/OSPSuite.Core/pull/1966

If you are OK with the Core PR, please merge it so I can also update the nuget packages here and have MoBi build. (I would just like to avoid having the packages with the letters suffix in the end if possible)

**Comment by msevestre** [comment]
done in core

**Comment by georgeDaskalakis** [comment]
@msevestre I am merging this with the updated packages :)

**Pull request merged**